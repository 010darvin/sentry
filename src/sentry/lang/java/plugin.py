from __future__ import absolute_import

from sentry.plugins import Plugin2
from sentry.stacktraces import StacktraceProcessor


FRAME_CACHE_VERSION = 1


class JavaStacktraceProcessor(StacktraceProcessor):

    def __init__(self, *args, **kwargs):
        StacktraceProcessor.__init__(self, *args, **kwargs)
        debug_meta = self.data.get('debug_meta')
        if debug_meta:
            self.available = True
            self.debug_meta = debug_meta
        else:
            self.available = False

    def handles_frame(self, frame, stacktrace_info):
        platform = frame.get('platform') or self.data.get('platform')
        return (
            platform == 'java' and
            self.available and
            'function' in frame
        )

    def preprocess_frame(self, processable_frame):
        pass

    def preprocess_step(self, processing_task):
        if not self.available:
            return False

    def process_frame(self, processable_frame, processing_task):
        pass


class JavaPlugin(Plugin2):
    can_disable = False

    def get_stacktrace_processors(self, data, stacktrace_infos,
                                  platforms, **kwargs):
        if 'java' in platforms:
            return [JavaStacktraceProcessor]
