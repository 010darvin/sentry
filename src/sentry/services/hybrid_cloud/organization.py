import dataclasses
from abc import abstractmethod
from dataclasses import dataclass, field
from enum import IntEnum
from typing import Iterable, List, Mapping, Optional, Sequence

from sentry.models import (
    Organization,
    OrganizationMember,
    OrganizationStatus,
    Project,
    ProjectStatus,
    ProjectTeam,
    Team,
    TeamStatus,
)
from sentry.roles.manager import TeamRole
from sentry.services.hybrid_cloud import (
    CreateStubFromBase,
    InterfaceWithLifecycle,
    logger,
    silo_mode_delegation,
)
from sentry.silo import SiloMode


class ApiTeamStatus(IntEnum):
    VISIBLE = TeamStatus.VISIBLE
    PENDING_DELETION = TeamStatus.PENDING_DELETION
    DELETION_IN_PROGRESS = TeamStatus.DELETION_IN_PROGRESS


class ApiProjectStatus(IntEnum):
    VISIBLE = ProjectStatus.VISIBLE
    HIDDEN = ProjectStatus.HIDDEN
    PENDING_DELETION = ProjectStatus.PENDING_DELETION
    DELETION_IN_PROGRESS = ProjectStatus.DELETION_IN_PROGRESS


@dataclass
class ApiTeam:
    id: int = -1
    status: ApiTeamStatus = ApiTeamStatus.VISIBLE
    organization_id: int = -1
    slug: str = ""


@dataclass
class ApiTeamMember:
    id: int = -1
    is_active: bool = False
    role: Optional[TeamRole] = None
    project_ids: List[int] = field(default_factory=list)
    scopes: List[str] = field(default_factory=list)
    team: ApiTeam = field(default_factory=ApiTeam)


@dataclass
class ApiProject:
    id: int = -1
    slug: str = ""
    name: str = ""
    organization_id: int = -1
    status: ApiProjectStatus = ApiProjectStatus.VISIBLE


@dataclass
class ApiOrganizationMember:
    id: int = -1
    organization_id: int = -1
    # This can be null when the user is deleted.
    user_id: Optional[int] = None
    teams: List[ApiTeamMember] = field(default_factory=list)
    role: str = ""
    projects: List[ApiProject] = field(default_factory=list)
    scopes: List[str] = field(default_factory=list)


@dataclass
class ApiOrganizationFlags:
    allow_joinleave: bool = False
    enhanced_privacy: bool = False
    disable_shared_issues: bool = False
    early_adopter: bool = False
    require_2fa: bool = False
    disable_new_visibility_features: bool = False
    require_email_verification: bool = False


@dataclass
class ApiOrganization:
    slug: str = ""
    id: int = -1
    # exists if and only if the organization was queried with a user_id context, and that user_id
    # was confirmed to be a member.
    member: Optional[ApiOrganizationMember] = None
    name: str = ""

    # Represents the full set of teams and projects associated with the org.
    teams: List[ApiTeam] = field(default_factory=list)
    projects: List[ApiProject] = field(default_factory=list)

    flags: ApiOrganizationFlags = field(default_factory=lambda: ApiOrganizationFlags())


class OrganizationService(InterfaceWithLifecycle):
    @abstractmethod
    def get_organization_by_id(
        self, *, id: int, user_id: Optional[int]
    ) -> Optional[ApiOrganization]:
        """
        Fetches the organization, team, and project data given by an organization id.  When user_id is provided,
        membership data related to that user from the organization is also given in the response.
        """
        pass

    @abstractmethod
    def get_organizations(
        self, user_id: Optional[int], scope: Optional[str], only_visible: bool
    ) -> List[ApiOrganization]:
        """
        When user_id is set, returns all organization and membership data associated with that user id given
        a scope and visibility requirement.  When user_id is None, provides all organizations across the entire
        system.
        """
        pass

    @abstractmethod
    def check_membership_by_email(
        self, organization_id: int, email: str
    ) -> Optional[ApiOrganizationMember]:
        """
        Used to look up an organization membership by an email
        """
        pass

    @abstractmethod
    def check_organization_by_slug(self, *, slug: str, only_visible: bool) -> Optional[int]:
        """
        If exists and matches the only_visible requirement, returns an organization's id by the slug.
        When allow_stale is true, returns
        """
        pass

    def get_organization_by_slug(
        self, *, user_id: Optional[int], slug: str, only_visible: bool
    ) -> Optional[ApiOrganization]:
        org_id = self.check_organization_by_slug(slug=slug, only_visible=only_visible)
        if org_id is None:
            return None

        return self.get_organization_by_id(id=org_id, user_id=user_id)

    def _serialize_member(
        self,
        member: OrganizationMember,
        team_to_project_ids: Optional[Mapping[int, int]] = None,
        projects: Optional[List[Project]] = None,
    ) -> ApiOrganizationMember:
        api_member = ApiOrganizationMember(
            id=member.id,
            organization_id=member.organization_id,
            user_id=member.user.id if member.user is not None else None,
            role=member.role,
            scopes=list(member.get_scopes()),
        )

        api_member.teams = [self._serialize_team(t) for t in member.teams]

        if projects is None:
            projects = Project.objects.filter(
                projectteam__team__organizationmemberteam__organizationmember=member
            )
        else:
            project_ids = {team_to_project_ids[t.id] for t in member.teams}
            projects = [p for p in projects if p.id in project_ids]

        api_member.projects = [self._serialize_project(p) for p in projects]

        return api_member

    def _serialize_flags(self, org: Organization) -> ApiOrganizationFlags:
        result = ApiOrganizationFlags()
        for f in dataclasses.fields(result):
            setattr(f, f.name, getattr(org.flags, f.name))
        return result

    def _serialize_team(self, team: Team) -> ApiTeam:
        return ApiTeam(
            id=team.id,
            status=team.status,
            organization_id=team.organization_id,
            slug=team.slug,
        )

    def _serialize_project(self, project: Project) -> ApiProject:
        return ApiProject(
            id=project.id,
            slug=project.slug,
            name=project.name,
            organization_id=project.organization_id,
            status=project.status,
        )

    def _serialize_organization(
        self, org: Organization, membership: Iterable[OrganizationMember] = tuple()
    ) -> ApiOrganization:
        api_org: ApiOrganization = ApiOrganization(
            slug=org.slug,
            id=org.id,
            flags=self._serialize_flags(org),
            name=org.name,
        )

        teams = Team.objects.filter(organization_id=org.id)
        projects = Project.objects.filter(organization_id=org.id)
        team_to_project_ids = {
            t.team_id: t.project_id
            for t in ProjectTeam.objects.filter(project_id__in=[p.id for p in projects])
        }

        api_org.teams = [self._serialize_team(t) for t in teams]
        api_org.projects = [self._serialize_project(p) for p in projects]

        for member in membership:
            if member.organization.id == org.id:
                api_org.member = self._serialize_member(member, team_to_project_ids, projects)
                break

        return api_org


class DatabaseBackedOrganizationService(OrganizationService):
    def get_organization_by_id(
        self, *, id: int, user_id: Optional[int]
    ) -> Optional[ApiOrganization]:
        membership: Sequence[OrganizationMember]
        if user_id is not None:
            membership = OrganizationMember.objects.filter(user_id=user_id).prefetch_related(
                "teams"
            )
        else:
            membership = []

        try:
            org = Organization.objects.get(id)
        except Organization.DoesNotExist:
            return None

        return self._serialize_organization(org, membership)

    def check_membership_by_email(
        self, organization_id: int, email: str
    ) -> Optional[ApiOrganizationMember]:
        try:
            member = OrganizationMember.objects.get(organization_id=organization_id, email=email)
        except OrganizationMember.DoesNotExist:
            return None

        return self._serialize_member(member)

    def check_organization_by_slug(self, *, slug: str, only_visible: bool) -> Optional[int]:
        try:
            org = Organization.objects.get(slug=slug)
            if only_visible and org.status != OrganizationStatus.VISIBLE:
                raise Organization.DoesNotExist
            return org.id
        except Organization.DoesNotExist:
            logger.info("Organization by slug [%s] not found", slug)

        return None

    def close(self) -> None:
        pass

    def get_organizations(
        self, user_id: Optional[int], scope: Optional[str], only_visible: bool
    ) -> List[ApiOrganization]:
        if user_id is None:
            return []
        organizations = self._query_organizations(user_id, scope, only_visible)
        membership = OrganizationMember.objects.filter(user_id=user_id)
        return [self._serialize_organization(o, membership) for o in organizations]

    def _query_organizations(
        self, user_id: int, scope: Optional[str], only_visible: bool
    ) -> List[Organization]:
        from django.conf import settings

        if settings.SENTRY_PUBLIC and scope is None:
            if only_visible:
                return list(Organization.objects.filter(status=OrganizationStatus.ACTIVE))
            else:
                return list(Organization.objects.filter())

        qs = OrganizationMember.objects.filter(user_id=user_id)

        qs = qs.select_related("organization")
        if only_visible:
            qs = qs.filter(organization__status=OrganizationStatus.ACTIVE)

        results = list(qs)

        if scope is not None:
            return [r.organization for r in results if scope in r.get_scopes()]

        return [r.organization for r in results]


StubOrganizationService = CreateStubFromBase(DatabaseBackedOrganizationService)
organization_service: OrganizationService = silo_mode_delegation(
    {
        SiloMode.MONOLITH: lambda: DatabaseBackedOrganizationService(),
        SiloMode.REGION: lambda: DatabaseBackedOrganizationService(),
        SiloMode.CONTROL: lambda: StubOrganizationService(),
    }
)
