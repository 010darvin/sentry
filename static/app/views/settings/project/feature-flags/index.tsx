import {Fragment, useEffect, useState} from 'react';
import styled from '@emotion/styled';
import isEqual from 'lodash/isEqual';

import {
  addErrorMessage,
  addLoadingMessage,
  addSuccessMessage,
} from 'sentry/actionCreators/indicator';
import {ModalRenderProps, openModal} from 'sentry/actionCreators/modal';
import Button from 'sentry/components/button';
import EmptyStateWarning from 'sentry/components/emptyStateWarning';
import FeatureBadge from 'sentry/components/featureBadge';
import CompactSelect from 'sentry/components/forms/compactSelect';
import ExternalLink from 'sentry/components/links/externalLink';
import SearchBar from 'sentry/components/searchBar';
import SentryDocumentTitle from 'sentry/components/sentryDocumentTitle';
import {IconAdd} from 'sentry/icons/iconAdd';
import {t, tct} from 'sentry/locale';
import ProjectStore from 'sentry/stores/projectsStore';
import space from 'sentry/styles/space';
import {Project} from 'sentry/types';
import {defined} from 'sentry/utils';
import handleXhrErrorResponse from 'sentry/utils/handleXhrErrorResponse';
import useApi from 'sentry/utils/useApi';
import useOrganization from 'sentry/utils/useOrganization';
import usePrevious from 'sentry/utils/usePrevious';
import SettingsPageHeader from 'sentry/views/settings/components/settingsPageHeader';
import TextBlock from 'sentry/views/settings/components/text/textBlock';
import PermissionAlert from 'sentry/views/settings/organization/permissionAlert';

import {FlagModal} from './modals/flagModal';
import {SegmentModal} from './modals/segmentModal';
import {Card} from './card';
import {Promo} from './promo';

type Props = ModalRenderProps & {
  project: Project;
};

export default function ProjectFeatureFlags({project}: Props) {
  const api = useApi();
  const organization = useOrganization();

  const currentFlags = project.featureFlags;
  const hasAccess = organization.access.includes('project:write');
  const disabled = !hasAccess;
  const previousFlags = usePrevious(currentFlags);

  const [flags, setFlags] = useState(currentFlags ?? {});
  const [query, setQuery] = useState('');
  const [status, setStatus] = useState(null);

  const showPromo = Object.keys(flags).length === 0;

  useEffect(() => {
    if (!isEqual(currentFlags, previousFlags)) {
      setFlags(currentFlags ?? {});
    }
  }, [currentFlags, previousFlags]);

  function handleAddFlag() {
    openModal(modalProps => (
      <FlagModal
        {...modalProps}
        organization={organization}
        project={project}
        flags={flags}
      />
    ));
  }

  function handleEditFlag(flagKey: string) {
    openModal(modalProps => (
      <FlagModal
        {...modalProps}
        organization={organization}
        project={project}
        flags={flags}
        flagKey={flagKey}
      />
    ));
  }

  async function handleDeleteFlag(flagKey: string) {
    const newFeatureFlags = {...flags};
    delete newFeatureFlags[flagKey];

    try {
      const result = await api.requestPromise(
        `/projects/${organization.slug}/${project.slug}/`,
        {
          method: 'PUT',
          data: {featureFlags: newFeatureFlags},
        }
      );
      ProjectStore.onUpdateSuccess(result);
      addSuccessMessage(t('Successfully deleted feature flag'));
    } catch (error) {
      const message = t('Unable to delete feature flag');
      handleXhrErrorResponse(message)(error);
      addErrorMessage(message);
    }
  }

  async function handleActivateToggle(flagKey: string) {
    const newFeatureFlags = {...flags};
    newFeatureFlags[flagKey].enabled = !newFeatureFlags[flagKey].enabled;

    addLoadingMessage();

    try {
      const result = await api.requestPromise(
        `/projects/${organization.slug}/${project.slug}/`,
        {
          method: 'PUT',
          data: {featureFlags: newFeatureFlags},
        }
      );

      ProjectStore.onUpdateSuccess(result);
      addSuccessMessage(t('Successfully updated the feature flag'));
    } catch (error) {
      const message = t('Unable to update the feature flag');
      handleXhrErrorResponse(message)(error);
      addErrorMessage(message);
    }
  }

  function handleAddSegment(flagKey: string) {
    openModal(modalProps => (
      <SegmentModal
        {...modalProps}
        organization={organization}
        project={project}
        flags={flags}
        flagKey={flagKey}
      />
    ));
  }

  function handleEditSegment(flagKey: string, index: number) {
    openModal(modalProps => (
      <SegmentModal
        {...modalProps}
        organization={organization}
        project={project}
        flags={flags}
        flagKey={flagKey}
        segmentIndex={index}
      />
    ));
  }

  async function handleDeleteSegment(flagKey: string, index: number) {
    const newEvaluations = [...flags[flagKey].evaluation];
    newEvaluations.splice(index, 1);

    const newFeatureFlags = {...flags};
    newFeatureFlags[flagKey].evaluation = newEvaluations;

    try {
      const result = await api.requestPromise(
        `/projects/${organization.slug}/${project.slug}/`,
        {
          method: 'PUT',
          data: {featureFlags: newFeatureFlags},
        }
      );

      ProjectStore.onUpdateSuccess(result);
      addSuccessMessage(t('Successfully deleted segment'));
    } catch (error) {
      const message = t('Unable to delete segment');
      handleXhrErrorResponse(message)(error);
      addErrorMessage(message);
    }
  }

  async function handleSortSegments(flagKey: string, segmentIds: string[]) {
    const sortedSegments = segmentIds
      .map(segmentId =>
        flags[flagKey].evaluation.find(segment => String(segment.id) === segmentId)
      )
      .filter(defined);

    const newFeatureFlags = {...flags};
    newFeatureFlags[flagKey].evaluation = sortedSegments;

    setFlags(newFeatureFlags);

    try {
      const result = await api.requestPromise(
        `/projects/${organization.slug}/${project.slug}/`,
        {
          method: 'PUT',
          data: {featureFlags: newFeatureFlags},
        }
      );
      ProjectStore.onUpdateSuccess(result);
      addSuccessMessage(t('Successfully sorted segments'));
    } catch (error) {
      setFlags(previousFlags ?? []);
      const message = t('Unable to sort segments');
      handleXhrErrorResponse(message)(error);
      addErrorMessage(message);
    }
  }

  const filteredFlags = Object.keys(flags).filter(key => {
    return (
      key.toLowerCase().includes(query.toLowerCase()) &&
      (status === null || flags[key].enabled === status)
    );
  });

  return (
    <SentryDocumentTitle title={t('Feature Flags')}>
      <Fragment>
        <SettingsPageHeader
          title={
            <Fragment>
              {t('Feature Flags')} <FeatureBadge type="beta" />
            </Fragment>
          }
          action={
            !showPromo && (
              <Button
                title={
                  disabled ? t('You do not have permission to add flags') : undefined
                }
                priority="primary"
                size="sm"
                icon={<IconAdd size="xs" isCircled />}
                onClick={handleAddFlag}
                disabled={disabled}
              >
                {t('Add Flag')}
              </Button>
            )
          }
        />
        <TextBlock>
          {tct(
            'Feature flags allow you to configure your code into different flavors by dynamically toggling certain functionality on and off. Learn more about feature flags in our [link:documentation].',
            {
              link: <ExternalLink href="" />,
            }
          )}
        </TextBlock>

        <PermissionAlert access={['project:write']} />

        {showPromo ? (
          <Promo hasAccess={hasAccess} onGetStarted={handleAddFlag} />
        ) : (
          <Content>
            <Filters>
              <SearchBar
                size="md"
                onChange={setQuery}
                query={query}
                placeholder={t('Search by flag name')}
              />
              <CompactSelect
                defaultValue={status}
                options={[
                  {value: null, label: t('All')},
                  {value: true, label: t('Active')},
                  {value: false, label: t('Inactive')},
                ]}
                value={status}
                onChange={value => setStatus(value.value)}
                triggerProps={{prefix: t('Status'), style: {width: '100%'}}}
              />
            </Filters>
            {!filteredFlags.length ? (
              <EmptyStateWarning>
                <p>{t('Sorry, no feature flag match your filters')}</p>
              </EmptyStateWarning>
            ) : (
              <div>
                {filteredFlags.map(filteredFlag => (
                  <Card
                    key={filteredFlag}
                    flagKey={filteredFlag}
                    kind={flags[filteredFlag].kind}
                    enabled={flags[filteredFlag].enabled}
                    description={flags[filteredFlag].description}
                    segments={flags[filteredFlag].evaluation}
                    onDelete={() => handleDeleteFlag(filteredFlag)}
                    onEdit={() => handleEditFlag(filteredFlag)}
                    onActivateToggle={() => handleActivateToggle(filteredFlag)}
                    onAddSegment={() => handleAddSegment(filteredFlag)}
                    onDeleteSegment={id => handleDeleteSegment(filteredFlag, id)}
                    onEditSegment={id => handleEditSegment(filteredFlag, id)}
                    hasAccess={hasAccess}
                    onSortSegments={({reorderedItems}) =>
                      handleSortSegments(filteredFlag, reorderedItems)
                    }
                  />
                ))}
              </div>
            )}
          </Content>
        )}
      </Fragment>
    </SentryDocumentTitle>
  );
}

const Content = styled('div')`
  display: grid;
  gap: ${space(2)};
`;

const Filters = styled('div')`
  display: grid;
  gap: ${space(1)};
  grid-template-columns: 1fr max-content;
`;
