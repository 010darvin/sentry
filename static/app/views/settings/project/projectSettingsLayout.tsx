import {cloneElement, isValidElement, useMemo} from 'react';
import {RouteComponentProps} from 'react-router';

import * as AppStoreConnectContext from 'sentry/components/projects/appStoreConnectContext';
import {Organization, Project} from 'sentry/types';
import useRouteAnalyticsParams from 'sentry/utils/useRouteAnalyticsParams';
import withOrganization from 'sentry/utils/withOrganization';
import ProjectContext from 'sentry/views/projects/projectContext';
import SettingsLayout from 'sentry/views/settings/components/settingsLayout';
import ProjectSettingsNavigation from 'sentry/views/settings/project/projectSettingsNavigation';

type Props = {
  children: React.ReactNode;
  organization: Organization;
} & RouteComponentProps<{orgId: string; projectId: string}, {}>;

type InnerProps = Props & {project: Project};

function InnerProjectSettingsLayout({
  params,
  routes,
  project,
  organization,
  children,
  ...props
}: InnerProps) {
  // set analytics params for route based analytics
  useRouteAnalyticsParams(
    // use memoized value to avoid re-rendering
    useMemo(
      () => ({
        project_id: project.id,
        project_platform: project.platform,
      }),
      [project.id, project.platform]
    )
  );
  return (
    <AppStoreConnectContext.Provider project={project} organization={organization}>
      <SettingsLayout
        params={params}
        routes={routes}
        {...props}
        renderNavigation={() => <ProjectSettingsNavigation organization={organization} />}
      >
        {children && isValidElement(children)
          ? cloneElement<any>(children, {organization, project})
          : children}
      </SettingsLayout>
    </AppStoreConnectContext.Provider>
  );
}

function ProjectSettingsLayout({params, ...props}: Props) {
  const {orgId, projectId} = params;

  return (
    <ProjectContext orgId={orgId} projectId={projectId}>
      {({project}) => <InnerProjectSettingsLayout {...{params, project, ...props}} />}
    </ProjectContext>
  );
}

export default withOrganization(ProjectSettingsLayout);
