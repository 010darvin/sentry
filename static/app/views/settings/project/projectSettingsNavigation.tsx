import {useContext} from 'react';

import AppStoreConnectContext from 'sentry/components/projects/appStoreConnectContext';
import {Organization, Project} from 'sentry/types';
import withProject from 'sentry/utils/withProject';
import SettingsNavigation from 'sentry/views/settings/components/settingsNavigation';
import getConfiguration from 'sentry/views/settings/project/navigationConfiguration';

type Props = {
  organization: Organization;
  project?: Project;
};

const ProjectSettingsNavigation = ({organization, project}: Props) => {
  const appStoreConnectContext = useContext(AppStoreConnectContext);

  const debugFilesNeedsReview = appStoreConnectContext
    ? Object.keys(appStoreConnectContext).some(
        // @ts-expect-error TS(2532) FIXME: Object is possibly 'undefined'.
        key => appStoreConnectContext[key].credentials.status === 'invalid'
      )
    : false;

  return (
    <SettingsNavigation
      navigationObjects={getConfiguration({project, organization, debugFilesNeedsReview})}
      access={new Set(organization.access)}
      features={new Set(organization.features)}
      organization={organization}
      project={project}
    />
  );
};

export default withProject(ProjectSettingsNavigation);
