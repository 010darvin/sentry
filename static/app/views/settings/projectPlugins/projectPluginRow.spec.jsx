import {Organization} from 'fixtures/js-stubs/organization';
import {Plugin} from 'fixtures/js-stubs/plugin';
import {Project} from 'fixtures/js-stubs/project';
import {routerContext} from 'fixtures/js-stubs/routerContext';

import {render, screen, userEvent} from 'sentry-test/reactTestingLibrary';

import ProjectPluginRow from 'sentry/views/settings/projectPlugins/projectPluginRow';

describe('ProjectPluginRow', function () {
  const plugin = Plugin();
  const org = Organization({access: ['project:write']});
  const project = Project();
  const params = {orgId: org.slug, projectId: project.slug};
  const routerContext = routerContext([{organization: org, project}]);

  it('renders', function () {
    const {container} = render(
      <ProjectPluginRow {...params} {...plugin} project={project} />,
      {
        context: routerContext,
      }
    );

    expect(container).toSnapshot();
  });

  it('calls `onChange` when clicked', function () {
    const onChange = jest.fn();

    render(
      <ProjectPluginRow {...params} {...plugin} onChange={onChange} project={project} />,
      {context: routerContext}
    );

    userEvent.click(screen.getByRole('checkbox'));

    expect(onChange).toHaveBeenCalledWith('amazon-sqs', true);
  });

  it('can not enable/disable or configure plugin without `project:write`', function () {
    const onChange = jest.fn();

    render(
      <ProjectPluginRow {...params} {...plugin} onChange={onChange} project={project} />,
      {
        context: routerContext([{organization: Organization({access: []})}]),
      }
    );

    userEvent.click(screen.getByRole('checkbox'));

    expect(onChange).not.toHaveBeenCalled();
  });
});
