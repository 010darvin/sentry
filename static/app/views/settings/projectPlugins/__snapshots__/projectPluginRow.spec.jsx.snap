// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectPluginRow renders 1`] = `
<div>
  <div
    class="amazon-sqs css-119mele-PluginItem evund0f5"
  >
    <div
      class="css-10f9tt3-PluginInfo evund0f3"
    >
      <div
        class="css-t7m0sj-PluginIcon-StyledPluginIcon evund0f1"
        size="48"
      />
      <div
        class="css-1yqqovl-PluginDescription evund0f4"
      >
        <div
          class="css-13njfqo-PluginName evund0f2"
        >
          Amazon SQS 
          <span
            class="css-16i2hsm-Version evund0f0"
          >
            v8.23.0.dev0
          </span>
        </div>
        <div>
          <a
            class="css-1e7h2lz-Anchor-grayText ehytxfi0"
            href="https://github.com/getsentry/sentry"
            rel="noreferrer noopener"
            target="_blank"
          >
            Sentry Team
          </a>
          <span>
             
            Â·
             
            <a
              class="css-19t5w00-Link-grayText e1ycxor00"
              href="amazon-sqs"
            >
              Configure plugin
            </a>
          </span>
        </div>
      </div>
    </div>
    <button
      aria-checked="false"
      class="css-1lixgqp-SwitchButton ep74ita1"
      data-test-id="switch"
      role="checkbox"
      type="button"
    >
      <span
        class="css-1vgpr18-Toggle ep74ita0"
      />
    </button>
  </div>
</div>
`;
